File: common/src/main/kotlin/net/impleri/playerskills/skills/numeric/NumericSkill.kt
package net.impleri.playerskills.skills.numeric

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.TeamMode
import net.minecraft.resources.ResourceLocation

open class NumericSkill(
  name: ResourceLocation,
  value: Double?,
  description: String?,
  options: List<Double>?,
  changesAllowed: Int?,
  teamMode: TeamMode?,
  notify: Boolean?,
  notifyKey: String?,
) : Skill<Double>(
  name,
  NumericSkillType.NAME,
  value,
  description,
  options ?: ArrayList(),
  changesAllowed ?: UNLIMITED_CHANGES,
  teamMode ?: TeamMode.off(),
  notify ?: false,
  notifyKey,
) {
  // @TODO: Expose for modification
  val step = 1.0

  override fun copy(value: Double?, changesAllowed: Int): Skill<Double> {
    return NumericSkill(name, value, description, options, changesAllowed, teamMode, notify, notifyKey)
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/numeric/NumericSkill.scala
package net.impleri.playerskills.skills.numeric

import net.impleri.playerskills.api.skills.ChangeableSkillOps
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.TeamMode
import net.minecraft.resources.ResourceLocation

case class NumericSkill(
  override val name: ResourceLocation,
  override val value: Option[Double] = None,
  override val description: Option[String] = None,
  override val options: List[Double] = List.empty,
  override val changesAllowed: Int = Skill.UNLIMITED_CHANGES,
  override val teamMode: TeamMode = TeamMode.Off(),
  override val announceChange: Boolean = false,
  override val notifyKey: Option[String] = None,
  step: Double = NumericSkill.DefaultStep,
) extends Skill[Double] with ChangeableSkillOps[Double, NumericSkill] {
  override val skillType: ResourceLocation = NumericSkillType.NAME

  override protected[playerskills] def mutate(value: Option[Double], changesAllowed: Int): NumericSkill = {
    copy(value = value, changesAllowed = changesAllowed)
  }
}

object NumericSkill {
  val DefaultStep: Double = 1.0
}
